zebra-state/src/service/check/nullifier.rs

```diff
@@ -0,0 +1,119 @@
+//! Sprout and Sapling and Orchard nullifiers are considered disjoint,
+//! even if they have the same bit pattern."
+//!
+//! https://zips.z.cash/protocol/protocol.pdf#nullifierset
+
+use std::collections::HashSet;
+
+use tracing::trace;
+
+use crate::{
+    error::DuplicateNullifierError, service::finalized_state::FinalizedState, PreparedBlock,
+    ValidateContextError,
+};
+
+/// Reject double-spends of nullifers:
+/// - one from this [`PreparedBlock`], and the other already committed to the [`FinalizedState`].
+///
+/// (Duplicate non-finalized nullifiers are rejected during the chain update,
+/// see [`add_to_non_finalized_chain_unique`] for details.)
+///
+/// "A transaction is not valid if it would have added a nullifier
+/// to the nullifier set that already exists in the set"
+///
+/// https://zips.z.cash/protocol/protocol.pdf#commitmentsandnullifiers
+#[tracing::instrument(skip(prepared, finalized_state))]
+pub(crate) fn no_duplicates_in_finalized_chain(
+    prepared: &PreparedBlock,
+    finalized_state: &FinalizedState,
+) -> Result<(), ValidateContextError> {
+    for nullifier in prepared.block.sprout_nullifiers() {
+        if finalized_state.contains_sprout_nullifier(nullifier) {
+            Err(nullifier.duplicate_nullifier_error(true))?;
+        }
+    }
+
+    // TODO: sapling and orchard nullifiers (#2231)
+
+    Ok(())
+}
+
+/// Reject double-spends of nullifers:
+/// - both within the same [`JoinSplit`] (sprout only),
+/// - from different [`JoinSplit`]s, [`sapling::Spend`]s or [`Action`]s
+///   in this [`Transaction`]'s shielded data, or
+/// - one from this shielded data, and another from:
+///   - a previous transaction in this [`Block`], or
+///   - a previous block in this non-finalized [`Chain`].
+///
+/// (Duplicate finalized nullifiers are rejected during service contextual validation,
+/// see [`no_duplicates_in_finalized_chain`] for details.)
+///
+/// "A transaction is not valid if it would have added a nullifier
+/// to the nullifier set that already exists in the set"
+///
+/// https://zips.z.cash/protocol/protocol.pdf#commitmentsandnullifiers
+#[tracing::instrument(skip(chain_nullifiers, shielded_data_nullifiers))]
+pub(crate) fn add_to_non_finalized_chain_unique<'block, NullifierT>(
+    chain_nullifiers: &mut HashSet<NullifierT>,
+    shielded_data_nullifiers: impl IntoIterator<Item = &'block NullifierT>,
+) -> Result<(), ValidateContextError>
+where
+    NullifierT: DuplicateNullifierError + Copy + std::fmt::Debug + Eq + std::hash::Hash + 'block,
+{
+    for nullifier in shielded_data_nullifiers.into_iter() {
+        trace!(?nullifier, "adding nullifier");
+
+        // reject the nullifier if it is already present in this non-finalized chain
+        if !chain_nullifiers.insert(*nullifier) {
+            Err(nullifier.duplicate_nullifier_error(false))?;
+        }
+    }
+
+    // TODO:  test that the chain's nullifiers are not modified on error (this PR)
```

Comment:
Oops I forgot this test, and I'm out of time today. But it's actually pretty important, so I'll do it first thing tomorrow.

Reply:
The change was big, and this PR is almost ready to merge, so I split it into PR #2495.